### PR TITLE
[Feature] :: Global datasources

### LINKED ISSUE #1043: Support sharing database queries between different apps
**Hope to support sharing database connections between different pages and apps.**

**For example**

**Created a PostgreSQL connection on the pg_source page**
image
**I hope that I can choose to use the PostgreSQL that has been created.**
image

thanks

### PR DESCRIPTION
**Pull request by vjaris42**
fixes #1043

### CODE REVIEW

File: server/data-migrations/1675368628726-AddAppVersionToDataQueries.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        referencedColumnNames: ['id'],
+        referencedTableName: 'app_versions',
+      })
+    );
+
+    const entityManager = queryRunner.manager;
+    const organizations = await entityManager.find(Organization, {
+      select: ['id', 'name'],
+    });
+
```
Comment (gsmithun4) [inline server/data-migrations/1675368628726-AddAppVersionToDataQueries.ts:31]: This can be done
- get distinct of app_version_ids - A
- get a map query id vs app_version_id - B
- iterate A and get query ids under specific app_version_id from B - C
- update all query ids in C with specific app_version_id

A - select distinct(data_sources.app_version_id) from data_queries inner join data_sources on data_queries.data_source_id = data_sources.id
B - select data_queries.id, data_sources.app_version_id from data_queries inner join data_sources on data_queries.data_source_id = data_sources.id

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      select: ['id', 'name'],
+    });
+
+    await this.dropForeignKey('app_environments', 'app_version_id', queryRunner);
+    await queryRunner.query('alter table app_environments alter column app_version_id drop not null');
+
+    //Insert new environments under workspace
+    const newMappingForEnvironments = {};
+    for (const org of organizations) {
+      for (const { name, isDefault } of defaultAppEnvironments) {
```
Comment (gsmithun4) [inline server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts:19]:
```
 for (const org of organizations) {
      for (const { name, isDefault } of defaultAppEnvironments) {
```
this two loops are repeating 3 times, shall we combine three under one?

File: server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    const oldMappingForEnvironments = {};
+    for (const org of organizations) {
+      for (const { name } of defaultAppEnvironments) {
+        const envs = await queryRunner.query(
+          `select app_environments.id from app_versions inner join apps on apps.organization_id = $1
+                 inner join app_environments on app_environments.app_version_id = app_versions.id
+                  where app_versions.app_id = apps.id and app_environments.name=$2`,
+          [org.id, name]
+        );
+        console.log({ envs });
```
Comment (gsmithun4) [inline server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts:44]: remove console.logs
add logs which gives some info on each steps

File: server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        referencedTableName: 'organizations',
+        onDelete: 'CASCADE',
+      })
+    );
+
+    //Drop app_version_id column as it is no longer needed
+    await queryRunner.dropColumn('app_environments', 'app_version_id');
+  }
+
+  private async dropForeignKey(tableName: string, columnName: string, queryRunner) {
```
Comment (gsmithun4) [inline server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts:84]: this can be moved to a util file and re-used

File: server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
 import { MigrationInterface, QueryRunner, TableColumn, TableForeignKey } from 'typeorm';
 
-export class AddAppVersionIdColumnToDataQueries1639645243080 implements MigrationInterface {
```
Comment (gsmithun4) [inline server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts:3]: this file changes should be reverted and new migration files should be created

File: server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                   ?.join()})`,
+          [newMappingForEnvironments[name][org.id]]
+        );
+      }
+    }
+
+    //Delete old app_environments which are not under organizations
+    await entityManager.delete(AppEnvironment, { organizationId: null });
+
+    //Add Foreign key and delete constraints on organization_id column
```
Comment (gsmithun4) [inline server/data-migrations/1675844361118-MigrateEnvironmentsUnderWorkspace.ts:69]: Shall we move this to a new file?
not null constrain to `organization_id` should be added

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: server/data-migrations/1669054493160-moveDataSourceOptionsToEnvironment.ts [hunk 7]
```diff
@@ -31,7 +31,7 @@ export class moveDataSourceOptionsToEnvironment1669054493160 implements Migratio
         entityManager.create(AppEnvironment, {
           name,
           isDefault,
-          appVersionId: appVersion.id,
+          organizationId: appVersion.id,
```
Comment (gsmithun4) [inline server/data-migrations/1669054493160-moveDataSourceOptionsToEnvironment.ts:34]: why 
```
organizationId: appVersion.id,
```
?

File: server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts (hunk 5 above)
Comment (gsmithun4) [reply server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts:3]: please take a look

File: server/src/controllers/global_data_sources.controller.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import { AppsAbilityFactory } from 'src/modules/casl/abilities/apps-ability.factory';
+import { GlobalDataSourceAbilityFactory } from 'src/modules/casl/abilities/global-datasource-ability.factory';
+import { DataQueriesService } from '@services/data_queries.service';
+import { GlobalDataSourcesService } from '@services/global_data_sources.service';
+import { AuthorizeDataSourceOauthDto, CreateDataSourceDto, UpdateDataSourceDto } from '@dto/data-source.dto';
+import { decode } from 'js-base64';
+import { User } from 'src/decorators/user.decorator';
+import { DataSource } from 'src/entities/data_source.entity';
+
+@Controller('v2/data_sources')
```
Comment (gsmithun4) [inline server/src/controllers/global_data_sources.controller.ts:25]:
```
@Controller({
  path: "data_sources",
  version: '2'
})
```

Comment (gsmithun4) [reply server/src/controllers/global_data_sources.controller.ts:25]: add this to main.ts
```
app.enableVersioning({
    type: VersioningType.URI,
    defaultVersion: VERSION_NEUTRAL
  });
```

File: server/src/modules/global_data_sources/global_data_sources.module.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import { Module } from '@nestjs/common';
```
Comment (gsmithun4) [inline server/src/modules/global_data_sources/global_data_sources.module.ts:1]: Do we need a new module for global data source?
shall we add the controller to existing data sources module?
/src/modules/data_sources/data_sources.module.ts
```
controllers: [DataSourcesController, GlobalDataSourcesController],
```

File: server/src/services/global_data_sources.service.ts [hunk 10]
```diff
@@ -0,0 +1,357 @@
+import allPlugins from '@tooljet/plugins/dist/server';
```
Comment (gsmithun4) [inline server/src/services/global_data_sources.service.ts:1]: shall we re-use data_sources.service?

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: frontend/package.json [hunk 11]
```diff
@@ -6,6 +6,7 @@
     "@dnd-kit/core": "^6.0.5",
     "@dnd-kit/sortable": "^7.0.1",
     "@dnd-kit/utilities": "^3.2.0",
+    "@radix-ui/colors": "^0.1.8",
```
Comment (gsmithun4) [inline frontend/package.json:9]: Is this changes related to this PR?

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: frontend/package.json (hunk 11 above)
Comment (vjaris42) [reply frontend/package.json:9]: Yes this is for the color styling that is being used in figma for new designs. This library is used in storybook by Kiran.

File: frontend/src/App/App.jsx [hunk 12]
```diff
@@ -88,7 +89,7 @@ class App extends React.Component {
     return (
       <Suspense fallback={null}>
         <BrowserRouter history={history} basename={window.public_config?.SUB_PATH || '/'}>
-          <div className={`main-wrapper ${darkMode ? 'theme-dark' : ''}`} data-cy="main-wrapper">
+          <div className={`main-wrapper ${darkMode ? 'theme-dark dark-theme' : ''}`}>
```
Comment (gsmithun4) [inline frontend/src/App/App.jsx:92]:
```
data-cy="main-wrapper"
```
should be added

File: server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts
Comment (gsmithun4) [inline server/migrations/1675243901985-AddAppVersionIdColumnToDataQueries.ts:1]: Shall we revert this, please create new migration if changes required

File: server/templates/customer-dashboard/definition.json
Comment (gsmithun4) [inline server/templates/customer-dashboard/definition.json:1]: please revert if not related to this PR

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: server/templates/customer-dashboard/definition.json
Comment (vjaris42) [reply server/templates/customer-dashboard/definition.json:1]: This is related to current implementation have confirmed the same with Sherfin. Had to do this since the relationships have been changed.

File: frontend/src/Editor/LeftSidebar/SidebarDatasources.jsx [hunk 13]
```diff
@@ -69,42 +91,76 @@ export const LeftSidebarDataSources = ({
...
+        )}
       </div>
     );
   };
 
   const popoverContent = (
     <LeftSidebarDataSources.Container
       darkMode={darkMode}
-      renderDataSource={renderDataSource}
+      RenderDataSource={RenderDataSource}
```
Comment (kavinvenkatachalam) [inline frontend/src/Editor/LeftSidebar/SidebarDatasources.jsx:161]: Can we change it to camelCase and call it with a function? Looks bit odd with `capitalisation`

**Review by kavinvenkatachalam — CHANGES_REQUESTED** [review]
@vjaris42 Suggested a minor change. Please take a look

**Review by gsmithun4 — APPROVED** [review]
